packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
  * This way stringified values can be compared.
  */
-export interface ContainerRuntimeMessage {
+export interface ContainerRuntimeMessage<TType, TContents> {
 	/** Type of the op, within the ContainerRuntime's domain */
-	type: ContainerMessageType;
+	type: TType;
 	/** Domain-specific contents, interpreted according to the type */
-	contents: any;
+	contents: TContents;
 	/** Info describing how to handle this op in case the type is unrecognized (default: fail to process) */
 	compatDetails?: IContainerRuntimeMessageCompatDetails;
 }
 
+export type ContainerRuntimeDataStoreOpMessage = ContainerRuntimeMessage<
+	ContainerMessageType.FluidDataStoreOp,
+	IEnvelope
+>;
+export type InboundContainerRuntimeAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Attach,
+	InboundAttachMessage
+>;
+export type OutboundContainerRuntimeAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Attach,
+	IAttachMessage
+>;
+export type ContainerRuntimeChunkedOpMessage = ContainerRuntimeMessage<
+	ContainerMessageType.ChunkedOp,
+	IChunkedOp
+>;
+export type ContainerRuntimeBlobAttachMessage = ContainerRuntimeMessage<
+	ContainerMessageType.BlobAttach,
+	undefined
+>;
+export type ContainerRuntimeRejoinMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Rejoin,
+	undefined
+>;
+export type ContainerRuntimeAliasMessage = ContainerRuntimeMessage<
+	ContainerMessageType.Alias,
+	IDataStoreAliasMessage
+>;
+export type LocalContainerRuntimeIdAllocationMessage = ContainerRuntimeMessage<
+	ContainerMessageType.IdAllocation,
+	IdCreationRangeWithStashedState
+>;
+export type InTransitContainerRuntimeIdAllocationMessage = ContainerRuntimeMessage<
+	ContainerMessageType.IdAllocation,
+	IdCreationRange & { stashedState?: never }
+>;
+export type UnknownContainerRuntimeMessage = ContainerRuntimeMessage<
+	UnknownContainerMessageType,
+	unknown
+>;
+
+/**
+ * A {@link ContainerRuntimeMessage} that is received from the server and will be processed by the container runtime.
+ */
+export type InboundContainerRuntimeMessage =
+	| ContainerRuntimeDataStoreOpMessage
+	| InboundContainerRuntimeAttachMessage
+	| ContainerRuntimeChunkedOpMessage
+	| ContainerRuntimeBlobAttachMessage
+	| ContainerRuntimeRejoinMessage
+	| ContainerRuntimeAliasMessage
+	| InTransitContainerRuntimeIdAllocationMessage
+	| UnknownContainerRuntimeMessage;
+
+/** A {@link ContainerRuntimeMessage} that has been generated by the container runtime but is not yet being sent to the server. */
+export type LocalContainerRuntimeMessage =
+	| ContainerRuntimeDataStoreOpMessage
+	| OutboundContainerRuntimeAttachMessage
+	| ContainerRuntimeChunkedOpMessage
+	| ContainerRuntimeBlobAttachMessage
+	| ContainerRuntimeRejoinMessage
+	| ContainerRuntimeAliasMessage
+	| LocalContainerRuntimeIdAllocationMessage
+	| UnknownContainerRuntimeMessage;
+
+/** A {@link ContainerRuntimeMessage} that is being sent to the server from the container runtime. */
```

Comment:
With the three similar-but-different types here, is there some kind of test we could write that would fail to build if these types diverged?  Like, each of these types must cover all the known `ContainerMessageType` values - so having a switch statement in a test file or something?  Just thinking out loud here.